### PR TITLE
Enable key ids usage for private keys

### PR DESCRIPTION
**Pull request by lsubash**
This patch enables usage of key id for private keys  in slimboot repo. Key ids are configured in
 BuildLoader and platform BoardConfig files.  SLIMBOOT_KEY_DIR is set to default folder outside
 sblopen. 
Generation of extrenal Keyhash OS key hash to be configured  for QEMU/CGL/APL with appropriate keys.

Update travis for private key generation. GenerateKeys.py is invoked before auromated build gets triggered.

### CODE REVIEW

File: BootloaderCorePkg/Tools/GenCapsuleFirmware.py [hunk 1]
```diff
@@ -481,7 +481,7 @@ def ValidateRegistryFormatGuid(Argument):
     parser = argparse.ArgumentParser()
 
     parser.add_argument('-p',  '--payload', nargs=2, action='append', type=str, required=True, help='Specify payload information including GUID, FileName')
-    parser.add_argument('-k',  '--priv_key', dest='PrivKey', type=str, required=True, help='Private RSA 2048 key in PEM format to sign image')
+    parser.add_argument('-k',  '--priv_key', dest='PrivKey', type=str, required=True, help='Key Id or Private RSA 2048 key in PEM format to sign image')
```
Comment (sagraw2) [inline BootloaderCorePkg/Tools/GenCapsuleFirmware.py:484]: Does this has to be 2048 bit key ?

Comment (lsubash) [reply BootloaderCorePkg/Tools/GenCapsuleFirmware.py]: It can be RSA2K/3K keys. Updated help section.

File: Platform/QemuBoardPkg/BoardConfig.py [hunk 2]
```diff
@@ -182,12 +182,22 @@ def GetPlatformDsc (self):
...
     def GetKeyHashList (self):
         # Define a set of new key used for different purposes
-        # The key is either public key PEM format or private key PEM format
+        # The key is either key id or public key PEM format or private key PEM format
         pub_key_list = [
           (
-            # Use a single test key
-            HASH_USAGE['PUBKEY_CFG_DATA'] | HASH_USAGE['PUBKEY_FWU'] | HASH_USAGE['PUBKEY_OS'] | HASH_USAGE['PUBKEY_CONT_DEF'],
-            'TestSigningPrivateKey.pem'
+            # Use a single test key Id CFGDATA_KEY_ID
```
Comment (sagraw2) [inline Platform/QemuBoardPkg/BoardConfig.py:188]: Update these comments accordingly:
Key for verifying Config data blob.

File: BuildLoader.py [hunk 3]
```diff
@@ -88,11 +88,15 @@ def prep_env ():
         print("Unsupported operating system !")
         sys.exit(1)
 
+    if 'SLIMBOOT_KEY_DIR' not in os.environ:
```
Comment (sagraw2) [inline BuildLoader.py:91]: I am trying to understand why is this needed?

File: Platform/QemuBoardPkg/Script/TestCases/firmware_update.py [hunk 4]
```diff
@@ -257,7 +257,7 @@ def main():
     cmd = [ sys.executable,
             'BootloaderCorePkg/Tools/GenCapsuleFirmware.py',
             '-p',  'BIOS', bios_img,
-            '-k',  'BootloaderCorePkg/Tools/Keys/TestSigningPrivateKey.pem',
+            '-k',  '../SlimbootKeyDir/FirmwareUpdateTestKey_Priv_RSA2048.pem',
```
Comment (sagraw2) [inline Platform/QemuBoardPkg/Script/TestCases/firmware_update.py:260]: Why is the default path for keys changed from BootloaderCorePkg/Tools/Keys to SlimbootKeyDir/ ?

File: BuildLoader.py (hunk 3 above)
Comment (lsubash) [reply BuildLoader.py:91]: While slimboot build, we could have per-checks for required env.
As we use key_id's it would be to add check as pre-build. Adding help note to generate keys need to be added here.

In standlaone tools + Key id - check in singlesing.py would be used.

File: Platform/QemuBoardPkg/Script/TestCases/firmware_update.py (hunk 4 above)
Comment (lsubash) [reply Platform/QemuBoardPkg/Script/TestCases/firmware_update.py:260]: BootloaderCorePkg/Tools/Key would would have keys.
Generated keys likely to be in ../SlimbootKeyDir/. 
Instead of path we can use FIRMWAREUPDATE_KEY_ID.  Due to env setting issues in travis builds i have used direct key path.

File: .travis.yml [hunk 5]
```diff
@@ -67,13 +67,13 @@ script:
   - |
     if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
       if [[ "$BUILD_TARGET" == "qemu" && "$BUILD_ARCH" == "" ]]; then
-        docker run --rm -v ${PWD}:/tmp/sbl -w /tmp/sbl --network=host sbl bash -c "python3 ./BuildLoader.py build ${BUILD_REL} ${BUILD_TARGET} && python3 ./Platform/QemuBoardPkg/Script/qemu_test.py";
+        docker run --rm -v ${PWD}:/tmp/sbl -w /tmp/sbl --network=host sbl bash -c "python3 BootloaderCorePkg/Tools/GenerateKeys.py -k ./../SlimbootKeyDir && python3 ./BuildLoader.py build ${BUILD_REL} ${BUILD_TARGET} && python3 ./Platform/QemuBoardPkg/Script/qemu_test.py";
```
Comment (mauricema) [inline .travis.yml:70]: ./../  should be ../

File: BootloaderCorePkg/Tools/BuildUtility.py [hunk 6]
```diff
@@ -859,6 +860,10 @@ def check_for_git():
     print_tool_version_info(cmd, version)
     return version
 
+def check_for_slimbootkeydir():
+    if not os.path.exists(os.environ.get('SLIMBOOT_KEY_DIR')):
+        raise Exception ("SLIMBOOT_KEY_DIR is not valid. Set correct SLIMBOOT_KEY_DIR path !!")
```
Comment (mauricema) [inline BootloaderCorePkg/Tools/BuildUtility.py:865]: let us use SBL_KEY_DIR for consistency.

File: BootloaderCorePkg/Tools/SingleSign.py [hunk 7]
```diff
@@ -111,13 +111,24 @@ def get_key_id (priv_key):
     else:
         return None
 
-def get_key_from_store (in_key):
+def get_key_dir ():
+    # Check Key store setting SLIMBOOT_KEY_DIR path
+    if 'SLIMBOOT_KEY_DIR' not in os.environ:
+        print_message_slimboot_key_store()
```
Comment (mauricema) [inline BootloaderCorePkg/Tools/SingleSign.py:117]: print_sbl_key_dir ()

File: BuildLoader.py [hunk 8]
```diff
@@ -122,14 +126,14 @@ def __init__(self, *args, **kwargs):
 
         # NOTE: Variables starting with '_' will not be exported to Platform.dsc
 
-        # Define default private key (PEM format) used to sign configuration data, firmware capsule and IAS image
-        key_dir = os.path.join('BootloaderCorePkg', 'Tools', 'Keys')
+        # Default key dir is set by SLIMBOOT_KEY_DIR. _KEY_DIR is set to NULL.
+        self._KEY_DIR = ''
         # Allow master key to be able to sign everything by default
         self._MASTER_KEY_USAGE      = HASH_USAGE['PUBKEY_CFG_DATA'] | HASH_USAGE['PUBKEY_FWU'] | HASH_USAGE['PUBKEY_OS'] | \
```
Comment (mauricema) [inline BuildLoader.py:132]: @sagraw2:
Since the master key can also be used to sign CFG_DATA/FWU/OS by default,   should we remove it to be consistent with   your documentation ?

File: .travis.yml (hunk 5 above)
Comment (lsubash) [reply .travis.yml:70]: Do you recommend SlimBootKeyDir to be internal to SblOpen in our travis builds?

Comment (mauricema) [reply .travis.yml:70]: ./../ is the same as ../  
So ./ is not required.
+-SblKeys
 |
+-SblOpen 
They should be at the same level.

File: Platform/QemuBoardPkg/Script/TestCases/firmware_update.py (hunk 4 above)
Comment (mauricema) [reply Platform/QemuBoardPkg/Script/TestCases/firmware_update.py:260]: Maybe just call "SblKeys"  for the dir ?

File: .travis.yml (hunk 5 above)
Comment (lsubash) [reply .travis.yml:70]: Thanks Maurice.
 I have mis interpreted earlier response for  ../

File: BuildLoader.py (hunk 8 above)
Comment (sagraw2) [reply BuildLoader.py:132]: Yes. That would align our 'default' implementation with the documentation.

File: .travis.yml [hunk 9]
```diff
@@ -9,44 +9,44 @@ matrix:
     - name: "QEMU GCC x86 D"
       os: linux
       language: python
-      env: BUILD_TARGET=qemu BUILD_REL="" BUILD_ARCH=""
+      env: BUILD_TARGET=qemu BUILD_REL="" BUILD_ARCH=""  SBL_KEY_DIR="/tmp/SblKeys"
```
Comment (lsubash) [inline .travis.yml:12]: Added env setting SBL_KEY_DIR

File: BootloaderCorePkg/Tools/GenContainer.py [hunk 10]
```diff
@@ -578,7 +578,7 @@ def extract (self, name = '', file_path = ''):
 
             # create header entry
             auth_type_str = self.get_auth_type_str (self.header.auth_type)
-            key_file = 'TestSigningPrivateKey.pem' if auth_type_str.startswith('RSA') else ''
+            key_file = 'CONTAINER_KEY_ID' if auth_type_str.startswith('RSA') else ''
             alignment = self.header.alignment
```
Comment (lsubash) [inline BootloaderCorePkg/Tools/GenContainer.py:582]: Use CONTAINER_KEY_ID and CONTAINER_COMP_KEY_ID while extracting component and creating layout.

File: BootloaderCorePkg/Tools/GenerateKeys.py [hunk 11]
```diff
@@ -49,19 +49,56 @@ def signing_keys():
...
             cmd = '%s genrsa -F4 -out %s/%s_RSA%s.pem %s' % (openssl_path, key_dir, Key_file_name, key_size, key_size)
             run_process (cmd.split())
 
+    # Generate test public keys
+    GenerateRsaPubKeys (openssl_path, key_dir, key_size)
+    return
+
+# Generate RSA Public Keys based on required key size
+# User is required to replace the public keys with their original keys.
+def GenerateRsaPubKeys (openssl_path, key_dir, key_size):
```
Comment (lsubash) [inline BootloaderCorePkg/Tools/GenerateKeys.py:83]: Generating dummy/test public keys for OS image. User/customer needs to replace with appropriate keys.

File: Platform/QemuBoardPkg/BoardConfig.py (hunk 2 above)
Comment (lsubash) [reply Platform/QemuBoardPkg/BoardConfig.py]: Updated descriptions.

File: Platform/QemuBoardPkg/Script/TestCases/firmware_update.py (hunk 4 above)
Comment (lsubash) [reply Platform/QemuBoardPkg/Script/TestCases/firmware_update.py]: Yes. Changed SlimbootKeyDir to SblKey which is an default keys folder

File: BootloaderCorePkg/Tools/BuildUtility.py (hunk 6 above)
Comment (lsubash) [reply BootloaderCorePkg/Tools/BuildUtility.py]: Updated to use SBL_KEY_DIR

File: BootloaderCorePkg/Tools/SingleSign.py (hunk 7 above)
Comment (lsubash) [reply BootloaderCorePkg/Tools/SingleSign.py]: Updated.

File: BuildLoader.py (hunk 8 above)
Comment (lsubash) [reply BuildLoader.py]: Removed self._MASTER_KEY_USAGE. 
Pub key hashes would be defined in boardconfig.py.

**Comment by lsubash** [comment]
Updated PR with minor change,
In generatekeys.py we are keeping the test private key generated for OS key.
